### PR TITLE
Ddp 4548 open api to gae 2

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
## Context
Created newer PR as it got messy after the develop merge!

This is a branch off updated pending Java back-end PR to GAE.
Gets OpenAPI docs built in CircleCI on pushes to develop branch and deployed as a service to GAE.
Currently using the default URL assigned by GAE. A nicer URL possible with update to dispatch.
https://pepper-api-spec-dot-broad-ddp-dev.uc.r.appspot.com/

## Checklist

- [ X] I have labeled the type of changes involved using the `C-*` labels.
- [X ] I have assessed potential risks and labeled using the `R-*` labels.
- [X ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [X ] I have considered security and privacy, and added `I-*` labels as needed
- [ X] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [X ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?
Click on https://pepper-api-spec-dot-broad-ddp-dev.uc.r.appspot.com/

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [X ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [X ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ version: 2.1
 
 references:
   ci_support_image: &ci_support_image
-                       broadinstitute/study-server-build:2020-04-10G
+                      broadinstitute/study-server-build:2020-04-17A
```
Comment (MocanaAtBroad) [inline .circleci/config.yml]: Newer image that contains `npm` as it is needed for open api doc build.

File: .circleci/config.yml [hunk 2]
```diff
@@ -88,7 +90,11 @@ commands:
               DEPLOY_ENV=dev
             ;;
             # @todo remove when done
-              study-server-parallel-tests)
+            study-server-parallel-tests)
+              DEPLOY_ENV=dev
+            ;;
+            # @todo remove when done
+            DDP-4548_OpenAPIToGAE)
               DEPLOY_ENV=dev
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:119]: For development purposes.

File: .circleci/config.yml [hunk 3]
```diff
@@ -250,12 +256,24 @@ commands:
...
+    steps:
       - run:
           name: Delete previous versions of service
           command: |
             set +x
             # extract the service name from the GAE yaml file
-            CONFIG_FILE="<<parameters.deploy_dir>>/<<parameters.jar_base_name>>.yaml"
+            CONFIG_FILE="<<parameters.deploy_dir>>/<<parameters.deploy_config_base_name>>.yaml"
             SERVICE=$(grep -Eo -m1 "^service:\s*.*" "${CONFIG_FILE}" | awk '{print $2}')
             echo "The service name is *${SERVICE}*"
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:307]: Refactored deletion of old services. Gets used for java stuff too.

File: .circleci/config.yml [hunk 4]
```diff
@@ -419,7 +437,6 @@ commands:
                     $CHECKSTYLE_OPTION \
                     $SKIPTESTS_OPTION \
                     -Ditext.license=output-config/itextkey.xml \
-                    -Dddp.firecloudKeysDir=output-config/fc_keys \
                     -Dconfig.file=output-build-config/testing-inmemorydb.conf  \
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:468]: Sneak in this cleanup. firecloudKeysDir no longer used.

File: .circleci/config.yml [hunk 5]
```diff
@@ -567,6 +584,27 @@ jobs:
...
+      - set-deployment-env
+      - auth-gcp-service-account
+      - run:
+          name: Deploy docs website
+          command: |
+            set -u
+            gcloud app deploy --version="${SHORT_GIT_SHA}" --stop-previous-version --project "broad-ddp-${ENVIRONMENT}" deploy/app.yaml
+      - delete-old-service-versions:
+            deploy_dir: deploy
+
```
Comment (MocanaAtBroad) [inline .circleci/config.yml:670]: The main command that actually does the deployment.

File: .circleci/config.yml [hunk 6]
```diff
@@ -628,5 +666,14 @@ workflows:
...
 
+  deploy-api-docs-workflow:
+    jobs:
+      - deploy-docs-job:
+          filters:
+            branches:
+              only:
+                - develop
+                - DDP-4548_OpenAPIToGAE
+
```
Comment (MocanaAtBroad) [inline .circleci/config.yml]: Note that we are only targeting develop for publishing docs. We can setup a custom domain name for server if needed.

File: pepper-apis/docs/specification/build.sh [hunk 7]
```diff
@@ -40,8 +40,9 @@ build_docs() {
     rm -rf "${BUILD}" || true
     mkdir -p "${BUILD}"
 
-    $(npm bin)/swagger-cli bundle --outfile "${BUILD}/pepper.yml" "${SRC}/pepper.yml"
-    $(npm bin)/speccy lint --rules="rules.yml" "${BUILD}/pepper.yml"
+    $(npm bin)/swagger-cli bundle --type json --outfile "${BUILD}/pepper.json" "${SRC}/pepper.yml"
+    $(npm bin)/swagger-cli bundle --type yaml --outfile ${BUILD}/pepper.yaml  "${SRC}/pepper.yml"
+    $(npm bin)/speccy lint --rules="rules.yml" "${BUILD}/pepper.json"
```
Comment (MocanaAtBroad) [inline pepper-apis/docs/specification/build.sh:45]: Let's generate a yaml and json  formatted copies of the OpenAPI spec. (note that redoc works with the json version)

File: pepper-apis/docs/specification/deploy/app.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+handlers:
+  - url: /spec/(.*\.(yaml|json))
+    static_files: spec/\1
+    upload: spec/(.*\.(yaml|json))
+    secure: always
+
+  - url: /(.*)
+    static_files: index.html
+    upload: index.html
+    secure: always
```
Comment (MocanaAtBroad) [inline pepper-apis/docs/specification/deploy/app.yaml:18]: This is a purely static file site. Stuff to with scaling and python there just to keep GAE happy; no instances will be created.

File: pepper-apis/docs/specification/deploy/spec/pepper.json [hunk 9]
```diff
@@ -0,0 +1 @@
+../../build/pepper.json
```
Comment (MocanaAtBroad) [inline pepper-apis/docs/specification/deploy/spec/pepper.json:1]: Soft link to facilitate deployment.

File: pepper-apis/docs/specification/deploy/spec/pepper.yaml [hunk 10]
```diff
@@ -0,0 +1 @@
+../../build/pepper.yaml
```
Comment (MocanaAtBroad) [inline pepper-apis/docs/specification/deploy/spec/pepper.yaml:1]: Soft link to facilitate deployment.

**Review by bskinner — DISMISSED** [review]
Things look good to me 👍.

**Review by bskinner — APPROVED** [review]
👍

**Pull request merged**